### ISSUE TITLE
Numba 0.51: avoid subclassing NamedTuple in LiteralStrKeyDict

### ISSUE THREAD

**Issue by ehsantn** [numba/numba#6094]
## Reporting a bug

- [x] I have tried using the latest ~released~ release candidate version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [ ] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

`LiteralStrKeyDict` type is a dictionary and semantically different than a Tuple. Subclassing `NamedTuple` can cause issues, e.g. breaking all the `instance(t, types.BaseTuple)` calls. I understand that this approach makes implementation (e.g. lowering) easier, but is there a cleaner way to avoid breaking existing code?

**Comment by ehsantn** [comment]
CC @stuartarchibald

**Comment by stuartarchibald** [comment]
Thanks for the report. When I get some time I'll have a go at decoupling it. I presume this won't actually break existing code unless you elect to use the feature? or do you have an example of something that worked on 0.50 that now doesn't? Thanks again!

**Comment by ehsantn** [comment]
It breaks existing code since some functions can take both tuple and dictionaries. However, I was able to work around it since it requires just an extra type check. Some other issues along the same line:

- `LiteralStrKeyDict` is subclass of `Literal` but the values are not necessarily constant. We had code with this assumption.
- `types.Poison` and `interpreter._UNKNOWN_VALUE` (e.g. for dispatcher type) appear in nested literal values in some corner cases.

I'd be happy to discuss these case and help improve `LiteralStrKeyDict` since this is a very interesting and useful feature.

**Comment by stuartarchibald** [comment]
> It breaks existing code since some functions can take both tuple and dictionaries. 

Do you have an example of this please? The `LiteralStrKeyDict` type shouldn't be appearing for conditions that would have existed in 0.50 based code. This may be a bug.

> However, I was able to work around it since it requires just an extra type check.

Thanks, I'd anticipate that being the case. I think it's going to be possible to decouple these types from the classes they are borrowing from but I'm not sure of the LOE.

> Some other issues along the same line:
> 
>     * `LiteralStrKeyDict` is subclass of `Literal` but the values are not necessarily constant. We had code with this assumption.

I see. This feature is breaking new ground, we've never had any sort of container const-ness before (that wasn't emulated through analysis and transform in a pass) and I don't think there was a specific contract about what `Literal` should be. I'm not saying that what is present is right, just that that is what it is. I'm also unsure that `Literal` would imply constant value or the inverse, for example `None` is const but not literal. The reason these containers inherit from `Literal` is so that `literally` can be used to force re-dispatch and all that (incredibly complicated) mechanism can just be reused. Perhaps these containers need a `types.PartLiteral` to inherit from and that registering with all typing/`literally` mechanics so as to differentiate between `Literal` being a const thing and `Literal` being a const-like thing with some const-value semantics and some const-type semantics.

>     * `types.Poison` and `interpreter._UNKNOWN_VALUE` (e.g. for dispatcher type) appear in nested literal values in some corner cases.

Any chance you could provide reproducers please? That this is appearing isn't necessarily bad, `Poison` should prevent it being used.

> I'd be happy to discuss these case and help improve `LiteralStrKeyDict` since this is a very interesting and useful feature.

Thanks, aim is to be able to do stuff like:
```python
@njit
def my_function(data, config):
    tmp = data / np.linalg.norm(data, ord=config['normalize'])
    iv = config['initial_value']
    for i in tmp:
        iv += i
    return iv

@njit
def config_example(data):
    # pass a dictionary as configuration
    config_a = {'normalize': None, 'initial_value': 5}
    result_a = my_function(data, config_a)
    print(result_a)
    
    config_b = {'normalize': np.inf, 'initial_value': 10j}
    result_b = my_function(data, config_b)
    print(result_b)

config_example(np.arange(10.))
```

**Comment by ehsantn** [comment]
> > It breaks existing code since some functions can take both tuple and dictionaries.
> 
> Do you have an example of this please? The `LiteralStrKeyDict` type shouldn't be appearing for conditions that would have existed in 0.50 based code. This may be a bug.
> 

We have a separate infrastructure that passes constant values to specific functions (using a transform pass, not regular type inference). That's why some of these issues are exposed for us. Example function that takes both tuple and dictionary is pandas groupby `agg`.

> Perhaps these containers need a `types.PartLiteral` to inherit from and that registering with all typing/`literally` mechanics so as to differentiate between `Literal` being a const thing and `Literal` being a const-like thing with some const-value semantics and some const-type semantics.

`PartLiteral` sounds very good to me.

> 
> > ```
> > * `types.Poison` and `interpreter._UNKNOWN_VALUE` (e.g. for dispatcher type) appear in nested literal values in some corner cases.
> > ```
> 
> Any chance you could provide reproducers please? That this is appearing isn't necessarily bad, `Poison` should prevent it being used.
> 
Here is a simple reproducer for the `Poison` case. The `_UNKNOWN_VALUE` case was along the same lines but I can't reproduce it (will try again later).

```python
In [45]: @overload(f)
    ...: def f_overload(d):
    ...:     print(d)
    ...:     return lambda d: None
    ...:

In [46]: def f(d):
    ...:     return
    ...:

In [47]: @overload(f)
    ...: def f_overload(d):
    ...:     print(d)
    ...:     return lambda d: None
    ...:

In [48]: @numba.njit
    ...: def g():
    ...:     return f({"A": [lambda a: 2*a, "B"]})
    ...:
```

**Comment by stuartarchibald** [comment]
> Here is a simple reproducer for the Poison case. The _UNKNOWN_VALUE case was along the same lines but I can't reproduce it (will try again later).

Thanks, minor bug in type inference for the special case of a single entry dict ctor was the cause. https://github.com/numba/numba/pull/6109 fixes.

**Comment by stuartarchibald** [comment]
> LiteralStrKeyDict type is a dictionary and semantically different than a Tuple. Subclassing NamedTuple can cause issues, e.g. breaking all the instance(t, types.BaseTuple) calls. I understand that this approach makes implementation (e.g. lowering) easier, but is there a cleaner way to avoid breaking existing code?

As this would impact existing code using `isinstance` checks, this is fixed in #6111 and scheduled for 0.51 final.
